src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,320 @@
+            }
+            keepAliveResponseStreamObserver = new StreamObserver<LeaseKeepAliveResponse>() {
+                @Override
+                public void onNext(LeaseKeepAliveResponse leaseKeepAliveResponse) {
+                    processKeepAliveRespond(leaseKeepAliveResponse);
+                    Lease lease = keepAlives.get(leaseKeepAliveResponse.getID());
+                    if (lease != null && lease.isContainHandler()) {
+                        lease.getEtcdLeaseHandler().onKeepAliveRespond(leaseKeepAliveResponse);
+                    }
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+
+                }
+
+                @Override
+                public void onCompleted() {
+                }
+            };
+
+            initRequestStream(keepAliveResponseStreamObserver);
+
+            /**
+             * Start heartbeat schedule to keep alive leases and remove dead leases
+             */
+            if (keepAliveSchedule == null) {
+                keepAliveSchedule = Executors.newSingleThreadScheduledExecutor();
+            }
+            scheduledFuture = keepAliveSchedule.scheduleAtFixedRate(
+                    () -> {
+                        keepAliveExecutor();
+                        deadLineExecutor();
+                    }, 0, scanPeriod, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    /**
+     * It hints the state of the keep alive service.
+     *
+     * @return whether the keep alive service is running.
+     */
+    @Override
+    public boolean isKeepAliveServiceRunning() {
+        synchronized (this) {
+            return scheduledFuture != null && !scheduledFuture.isCancelled();
+        }
+    }
+
+    /**
+     * New a lease with ttl value
+     *
+     * @param ttl ttl value, unit seconds
+     * @return
+     */
+    @Override
+    public ListenableFuture<LeaseGrantResponse> grant(long ttl) {
+        LeaseGrantRequest leaseGrantRequest =
+                LeaseGrantRequest.newBuilder()
+                        .setTTL(ttl)
+                        .build();
+        return leaseFutureStub.leaseGrant(leaseGrantRequest);
+    }
+
+    /**
+     * revoke one lease and the key bind to this lease will be removed
+     *
+     * @param leaseId id of the lease to revoke
+     * @return
+     */
+    @Override
+    public ListenableFuture<LeaseRevokeResponse> revoke(long leaseId) {
+        LeaseRevokeRequest leaseRevokeRequest =
+                LeaseRevokeRequest.newBuilder()
+                        .setID(leaseId)
+                        .build();
+        return leaseFutureStub.leaseRevoke(leaseRevokeRequest);
+    }
+
+    /**
+     * keep alive one lease in background
+     *
+     * @param leaseId          id of lease to set handler
+     * @param etcdLeaseHandler the handler for the lease, this value can be null
+     */
+    @Override
+    public void keepAlive(long leaseId, EtcdLeaseHandler etcdLeaseHandler) {
```

Comment:
what if we call keepAlive concurrently? seems that we might end up with putting the same lease twice.